### ISSUE TITLE
Change precedence of Range

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#36514]
Below is an excerpt from an unedited version of today's LDM notes specifying a change to the syntax for Range.  This needs to be implemented and tested.

## Range

There are some syntactic ambiguities in the `Range` specification and implementation.  See https://github.com/dotnet/roslyn/issues/34483.  Principal among them are the fact that the specification gives no meaning to `a * .. b` but the compiler accepts it, producing a tree with precedence inversion.  One fix is to simply produce a syntax error in these cases.  However I believe this is likely to be confusing to the user.

These issues can be resolved by the following proposed precedence change:

1. `e1 .. e2` should be at a precedence between shift and additive (as currently specified).
2. `.. e` should be at unary precedence, like all the other prefix operators.
3. `e ..` should be at primary precedence, like all the other postfix operators.

Note that a single token look-ahead is required to distinguish between cases 1 and 3.  Nevertheless, this may result in some confusingly inconsistent parsing behavior.

``` c#
_ = a + b .. c; // (a + b) .. c
_ = a + b ..;   // a + (b ..)
```

This confusion could be eliminated by moving the precedence of the binary form from where it is (between *shift* and *additive*) to between *unary* and *multiplicative*:

``` c#
_ = a + b .. c; // a + (b .. c)
_ = a + b ..;   // a + (b ..)
```

This final syntactic proposal is summarized by the following grammar:

``` c#
multiplicative_expression
    : range_expression
    | multiplicative_expression '*' range_expression
    | multiplicative_expression '/' range_expression
    | multiplicative_expression '%' range_expression
    ;

range_expression
    : unary_expression
    | range_expression '..' unary_expression
    ;

unary_expression
    : prefix_range
    ; // plus all the other forms

prefix_range:
    : '..' unary_expression
    ;

primary_no_array_creation_expression:
    : primary_range
    ; // plus all the other forms

primary_range:
    : primary_expression '..'
    | '..'
    ;
```

There remain some small ambiguities (which admittedly are *caused* by using difference precedence levels).

- Is `.. ..` to be parsed as `.. (..)` or `(..) ..`?
- Is `.. .. e` to be parsed as `.. (.. e)` or `(..) .. (e)`?
- Is `.. e ..` to be parsed as `.. (e ..)` or `(.. e) ..`?
- Is `e1 .. .. e2` to be parsed as `e1 .. (.. e2)` or `(e1 ..) .. e2`?

Note that these are all semantic errors (as there is no form of the range operator that takes a range as an operand).

I suggest the former in all cases, which would permit limited look-ahead in the parser (look-ahead would be needed to distinguish the parses of `.. .. e` from `.. ..` for a different resolution).

An alternative approach is to simply move the precedence to the suggested level for all forms of `..` and give an error for precedence inversions, which should now be much less likely.

Anothed ambiguity arises from combination with other operators that have both a unary and binary form.  What is the meaning of `.. + a` or `.. - a`?
1. `.. (+ a)`
2. `(..) + a`

Do we have some systematic way of answering such questions?

Resolution: Make the proposed precedence changes per syntax above, and the disambiguaation rule is that if the token following `..` could start an expression, then it is the right-hand-side operand of the range operator.

**Comment by agocke** [comment]
@gafter I think there's conflict between your example `.. e ..` and the rule you propose. It looks like the resolution should be `(.. e) ..` since `e` can start an expression.

**Comment by gafter** [comment]
@agocke It is true that `e` can start an expression, but both possible parses treat that `..` as having a right-hand operand.  This ambiguity is resolved by the proposed precedence.  The other parse has a precedence inversion (`.. e` is a *prefix_expression* but the postfix version requires a *primary_expression* on the left-hand-side).

The disambiguation rule should be more precisely stated:
> the disambiguaation rule is that if the token following .. could start an expression, then it is the start of the right-hand-side operand of the range operator

**Comment by gafter** [comment]
An interesting case that illustrates how the disambiguation rule arises is

`a .. b ..`

This should parse as `a .. (b ..)`

`a .. b .. c`

This should parse as `(a .. b) .. c`

This is a case that shows how look-ahead is needed to determine the precedence of the second `..` to distinguish between these cases.

**Comment by CyrusNajmabadi** [comment]
> `a .. b .. c`

I thought we didn't want that to parse at all due to potentially wanting to reserve that as syntax for supplying a step amount for ranges.  Is that not the case?  Or are you sayin it would parse, but something else would restrict it from being usable?

**Comment by gafter** [comment]
@CyrusNajmabadi The `..` operator cannot be user-defined, and no built-in version of it takes a `Range` as an operand.  So it is a semantic error.

**Comment by CyrusNajmabadi** [comment]
> The .. operator cannot be user-defined, and no built-in version of it takes a Range as an operand. So it is a semantic error.

That seems slightly dangerous (but i imagine the LDM would be fine with the breaking change in the future if it ever came).  Namely, someone can provide their own source version of System.Range right?  Which could support this.  

I personally think it would be good to just explicitly carve this out in the spec/impl to ensure there's no issue in the future.  But i also acknowledge it's hugely unlikely to ever be a concern :)

**Comment by gafter** [comment]
> That seems slightly dangerous (but i imagine the LDM would be fine with the breaking change in the future if it ever came). Namely, someone can provide their own source version of System.Range right? Which could support this.

No.  The language/compiler defines the types accepted and returned by `..`.  Not some API.

**Comment by CyrusNajmabadi** [comment]
> No. The language/compiler defines the types accepted and returned by ..

Really?  I thought the idea of `x..y` was that it became calls to System.Range and System.Index as appropriate.  So if `x..y` becomes a call to `new System.Range(x, y)`, then wouldn't `x..y..z` become a call to `new System.Range(x, new System.Range(y, z))`?  Or is the language mandating that the specific constructor it must find must be of the form `Range(System.Index, System.Index)`?  If so, what would happen if there was an implicit conversion from Range to Index, making the above into `new System.Range(x, (Implicit conversion to Index)(new System.Range(y, z))`?

**Comment by CyrusNajmabadi** [comment]
@gafter I just tried this and this compiled with the version of VS i have:

```c#
namespace System
{
    public struct Range
    {
        public Range(Index i, Index j)
        {

        }

        public static implicit operator Index(Range r) => 0;
    }

    public struct Index
    {
        public Index(int i)
        {

        }

        public static implicit operator Index(int i)
        {
            return new Index(i);
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        var v = 1..(2..3);
    }
}
```

If precedence allows for `1..2..3` to be parsed as `1..(2..3)` then the sort of code i wrote would succeed at compiling.  This would make it a breaking change in the future ot make `x..y..z` into a `range with step` expression.  

So, my recommendation would be to have something that explicitly disallows this for now (perhaps semantically), until there is a decision on if we want `range with step` in the language in the future, and what syntax that would have. 

Thanks!

**Comment by gafter** [comment]
> If precedence allows for `1..2..3` to be parsed as `1..(2..3)`

It doesn't.  It is parsed as `(1..2)..3`.

> This would make it a breaking change in the future to make `x..y..z` into a range with step expression.

We can do that based on the current parse of `(x..y)..z`.  I don't think we will, but we could.

**Comment by CyrusNajmabadi** [comment]
> It doesn't. It is parsed as (1..2)..3.

Sorry, that's what i meant.  I'm simply saying that if we allow parsing `1..2..3` then code like the stuff i showed will be legal, and if we ever want to add step-ranges, tehn that would break this.

> We can do that based on the current parse of (x..y)..z. I don't think we will, but we could.

How?

If we allow parsing today of `x..y..z` (which my program demonstrates would compile), then we can't use that as the syntactic form for step-ranges without potentially breaking code.  I agree it's marginal, but it just seems safer to disallow `x..y..z` early on to prevent any concerns later.

**Comment by gafter** [comment]
> How?

By adding an operator `..` that takes a Range (without a specified step) on the left and an integer on the right, and returns a Range (with a step).

**Comment by agocke** [comment]
@gafter I think the change to postfix `..` may introduce a problem. By my reading, `-5..` would parse as `-(5..)` since postfix `..` is higher precedence than `-`. That looks wrong.

**Comment by gafter** [comment]
@agocke I think that's fine for a few reasons.  The range operator does not accept negative values.  Also, the "preferred" parse would be a semantic error, as the range type does not have a negation operator.  How would you want `await e..` to parse?

**Comment by agocke** [comment]
^ is also a unary operator, though

**Comment by gafter** [comment]
@agocke Hmmm, good point.  Do you have a proposal that won't result in precedence inversion?

**Comment by agocke** [comment]
@gafter What if the postfix operator is also unary precedence?

**Comment by gafter** [comment]
@agocke You'd have some precedence inversion issues.  For example you'd have the problem of interpreting `a.. .Start`  It looks like it should mean `(a..) .Start` and could have no other possible parse, but you cannot have a unary operator on the left of the member selection operator due to [precedence](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#operator-precedence-and-associativity).

**Comment by CyrusNajmabadi** [comment]
> For example you'd have the problem of interpreting `a.. .Start`

FWIW, i'd have no problem with this.  I have no expectation of the above being parsable.  I would want anyone writing that to need to write it as `(a..).Start` instead.  It would be much clearer and in-line with my expectations.

**Comment by jcouv** [comment]
@agocke Can we close this? Is this duplicate of https://github.com/dotnet/roslyn/issues/34483

**Comment by gafter** [comment]
@jcouv We need an issue to track the minor precedence changes to range agreed this week by LDM.
